### ISSUE TITLE
Fix Visual Editor to output Royale Jewel components instead of FlexJS

### ISSUE THREAD

**Issue by JustinProminic** [prominic/Moonshine-IDE#675]
It is time we address this... FlexJS should be removed, in favor of Royale Jewel component output.

The general goal is for someone to build a working Royale Jewel application within 5 minutes after installing Moonshine, and no more than 10 mouse clicks.

**Comment by piotrzarzycki21** [comment]
@JustinProminic I'm not quite sure what do you mean here. If you have in mind having another type of project in Visual Editor where user can create UI interfaces by drag/drop on surface area - same as we have for PrimeFaces etc. - This is rather huger tasks and it should be named differently probably.

**Comment by rat-moonshine** [comment]
I probably a little confused, @JustinProminic . I believe we decided to obsolete FlexJS long back in Moonshine. I don't see we have any support for FlexJS either, 

![image](https://user-images.githubusercontent.com/24370870/76054101-84bd8480-5f95-11ea-8b9c-93628e68285d.png)

Perhaps, you meant about Visual Editor support that we have for Flex and PrimeFaces?

I think I understood that you wants same mock-up support for Royale Jewel projects, too - whether you wants to remove any existing _type_ (Flex, PrimeFaces) and add Jewel, or it'll be just a new addition, that is not clear to me. I don't think we output for FlexJS through Visual Editor, though.

**Comment by JustinProminic** [comment]
@rat-moonshine @piotrzarzycki21 @JoelProminic 

Visual Editor project needs to support - in the long run (no particular order here):

1) PrimeFaces (current support is rudimentary but functional at least)

2) Notes/Domino forms (under development by @feather812002 Bing)

3) Royale Jewel UI component -- a replacement for what we had in FlexJS.  I did not realize we had ripped this out.  I thought the code was still pretty much in Visual Editor but the output format was only targeting FlexJS.   

and when Josh is ready:
4) FeathersUI for Haxe output, in various formats that FeathersUI will work (@joshtynjala )

These are the value added features that make Moonshine truly unique, and harken back to the Flash/Flex Builder days (but not attempting round-trip like it did).


So, if I misunderstood where we are at with the FlexJS code, we can talk about that on the next Wednesday meeting.  I was merely trying to lay out the roadmap for updating this feature so that those kicking the tires on Royale Jewel components can use the existing Visual Editor to create a very basic form without coding.

**Comment by JustinProminic** [comment]
Furthermore, we had an entire system we built a while ago that would take Notes/Domino forms and convert them to FlexJS.  I'm pretty sure the intermediate format was the Visual Editor XML.  So the flow was Notes/Domino DXL -> Visual Editor -> FlexJS.

@JoelProminic am I misremembering this?

**Comment by piotrzarzycki21** [comment]
> 3\. Royale Jewel UI component -- a replacement for what we had in FlexJS.  I did not realize we had ripped this out.  I thought the code was still pretty much in Visual Editor but the output format was only targeting FlexJS.

W have never had in Visual Editor output to FlexJS, unless I'm not aware of something and code is laying somewhere on the Prominic sight where I don't have access. As @rat-moonshine said we do have output in Visual Editor to Flex and PrimeFaces. Making Visual Editor for Royale Jewel is something new and unexplored yet.

**Comment by rat-moonshine** [comment]
We need to discuss on feasibility of this requirement since this issue is long opened and moving through milestones. As I understand FlexJS is now dominated by its next generation RoyaleJS. Although, it appears that Apache still distributes FlexJS source through their website. 

An issue like this is breadth of a milestone, do we want to keep this into our queued tasks?

**Comment by JustinProminic** [comment]
@feather812002 @rat-moonshine @JoelProminic @piotrzarzycki21 

I have just sent you  direct e-mail subject "Moonshine IDE -- Notes Client to FlexJS prototype -- Fix Visual Editor to output Royale Jewel components instead of FlexJS #675" 

dated:
11/09/2021 01:43 AM Chicago Time

Please read it as it has the details on our past efforts on this aspect of conversion.

**Comment by piotrzarzycki21** [comment]
> FlexJS is now dominates by its next generation RoyaleJS

@rat-moonshine no it's not "dominates" - It is exactly the same framework. We have changed project name to do not connect project fully with Flex.

**Comment by piotrzarzycki21** [comment]
> Although, it appears that Apache still distributes FlexJS source through their website.

Yep and it will be. Each release is archived so there is no way it will just disappear. I'm not sure where do you see that but it will stay for sure.

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 It could be, that FlexJS and Royale are same (because of the underlaying framework both uses) - but as a layman I see both are different products (I'm sure we also puzzled between using which products by their names), and suggested to use Royale over FlexJS.

I've seen the distribution at here: https://flex.apache.org/download-flexjs.html.

**Comment by JoelProminic** [comment]
In any case, I would expect the FlexJS code from our previous work to be obsolete at this point.  I think we will probably be better off using the templates from #704 as an example.

@feather812002, take a look at the [template](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/704#issuecomment-643181684) created by @piotrzarzycki21.  We will probably need some more examples to implement all of the DominoVisualEditor options, but this will be a good place to start.

**Comment by JoelProminic** [comment]
@feather812002 is getting ready to switch to this task, so I wanted to supply some useful references:
- Example project:  https://github.com/Moonshine-IDE/Moonshine-IDE/issues/704#issuecomment-643181684
- Templates from #704:  https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates
- Royale Documentation:  https://royale.apache.org/docs/
- Jewel and Royale examples:  https://royale.apache.org/tourdejewel/

**Comment by JoelProminic** [comment]
Unfortunately, I found that the example project was failing:

    : /Users/-user-/Downloads/RoyaleTabularCRUDTemplate/src/RoyaleTabularCRUDTemplate.mxml(12): col: 2 Error: Cannot parse a value of type 'org.apache.royale.core.IApplicationView' from ''.
    : 	<j:initialView>
    : 	^
    : /Users/-user-/Downloads/RoyaleTabularCRUDTemplate/src/RoyaleTabularCRUDTemplate.mxml(13): col: 3 Error: In initializer for 'j:initialView', type view.MainContent is not assignable to target type 'org.apache.royale.core.IApplicationView'.
    : 		<view:MainContent />
    : 		^
    : /Users/-user-/Downloads/RoyaleTabularCRUDTemplate/src/view/MainContent.mxml(2): col: 1 Error: This tag could not be resolved to an ActionScript class. It will be ignored.
    : <j:ApplicationResponsiveView xmlns:fx="http://ns.adobe.com/mxml/2009"
    : ^


I did some minimal fixes to resolve the compilation errors, but I see that I can't submit the forms now.  @rat-moonshine or @piotrzarzycki21, can you revisit this template when you have a chance.  We can discuss this in more detail in the meeting tomorrow.

[2021_11_16__RoyaleTabularCRUDTemplate_JoelProminic.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/7550512/2021_11_16__RoyaleTabularCRUDTemplate_JoelProminic.zip)

**Comment by piotrzarzycki21** [comment]
> Unfortunately, I found that the example project was failing:
> 
> ```
> : /Users/-user-/Downloads/RoyaleTabularCRUDTemplate/src/RoyaleTabularCRUDTemplate.mxml(12): col: 2 Error: Cannot parse a value of type 'org.apache.royale.core.IApplicationView' from ''.
> : 	<j:initialView>
> : 	^
> : /Users/-user-/Downloads/RoyaleTabularCRUDTemplate/src/RoyaleTabularCRUDTemplate.mxml(13): col: 3 Error: In initializer for 'j:initialView', type view.MainContent is not assignable to target type 'org.apache.royale.core.IApplicationView'.
> : 		<view:MainContent />
> : 		^
> : /Users/-user-/Downloads/RoyaleTabularCRUDTemplate/src/view/MainContent.mxml(2): col: 1 Error: This tag could not be resolved to an ActionScript class. It will be ignored.
> : <j:ApplicationResponsiveView xmlns:fx="http://ns.adobe.com/mxml/2009"
> : ^
> ```
> 
> I did some minimal fixes to resolve the compilation errors, but I see that I can't submit the forms now. @rat-moonshine or @piotrzarzycki21, can you revisit this template when you have a chance. We can discuss this in more detail in the meeting tomorrow.
> 
> [2021_11_16__RoyaleTabularCRUDTemplate_JoelProminic.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/7550512/2021_11_16__RoyaleTabularCRUDTemplate_JoelProminic.zip)

What version of Apache Royale did you use to build that example ? I just tried it and I'm able build with official release 0.9.8 and fresh nightly 0.9.9.

**Comment by rat-moonshine** [comment]
> What version of Apache Royale did you use to build that example ? I just tried it and I'm able build with official release 0.9.8 and fresh nightly 0.9.9.

Does that means this basic template-codes is still valid with latest Royale nightly, and without any break?

**Comment by piotrzarzycki21** [comment]
> > What version of Apache Royale did you use to build that example ? I just tried it and I'm able build with official release 0.9.8 and fresh nightly 0.9.9.
> 
> Does that means this basic template-codes is still valid with latest Royale nightly, and without any break?

It looks like it is.

**Comment by JoelProminic** [comment]
I meant to discuss this in the meeting today, but it is just as well that we missed it with how late the meeting ran.

I tested with 0.9.9.  I think this instance was a few weeks old though.

To be clear, I got the error on the [previous example here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/704#issuecomment-643181684).   I resolved the compilation errors with my [updated example](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/675#issuecomment-970808097).

However, I'm seeing that the submit button on the form doesn't do anything currently.  Does this need to call some external API?  I didn't notice any errors on the browser console in my test.

**Comment by JoelProminic** [comment]
To be clear, we need to generate the Royale interface from the Domino components.  The intended workflow is something like this:
1. User wants to convert a NSF database to Royale
2. Use Moonshine to launch NsfConverter in a Vagrant VM (#770, milestone v3.3.2)
3. Convert the database to intermediate XML
4. Use the intermediate XML to build a Domino Visual Editor project
5. Make any desired modifications to the forms in the mockup
6. Generate a Royale interface from the mockups (this issue)
7. Generate a backend application to support the Royale mockup
8. Open the interface and backend (in Vagrant?) so that the user can test the interface

We can add toRoyaleCode() methods to the Domino components and start generating simple Royale code for this.  I originally wanted to have a separate converter classes so that the layout and component interactions could be handled with more context, but it seems this is how the previous converter code was working.


I'd suggest adding components in roughly this order, so that we can start testing limited forms early.  Focus on the required properties first, and we'll add the optional features later as we continue to improve the design.
1. Basic Labels
2. Basic layout and paragraphs
3. Basic Text Fields
4. Basic implementations of other field types
5. Tables
6. Tabs
7. Sections
8. Fill in the more complicated properties from the above examples

**Comment by JoelProminic** [comment]
To clarify, for now I want to add this logic to the Domino Visual Editor project rather than creating a separate Royale Visual Editor project.  To generate the Royale project, add a new button called "Project > Export Apache Royale Project".

This will be a little confusing for users who want to create a Royale project directly, so we may rename projects or make other updates for this later.  For now, I just want this to be an extension of the Domino Visual Editor project so that we can avoid duplication.

**Comment by JoelProminic** [comment]
@feather812002 reported that Moonshine could generate the basic Royale project.  This was updated in the `features/issue_675_royale_generate_domino_visual_editor` branch (only on Moonshine-IDE currently).

I am getting build errors in Bamboo:

    build	07-Dec-2021 13:15:14	CompileWeb:
    build	07-Dec-2021 13:15:14	     [copy] Copying 8029 files to /Users/mmb/bamboo-agent-home/xml-data/build-dir/MS-BMOB-JOB1/Moonshine-IDE/ide/MoonshineWEBevolved/build/bin-debug/elements
    build	07-Dec-2021 13:15:20	     [echo] /Users/mmb/SDKs/Flex_4.16.1_AIR_33.1
    build	07-Dec-2021 13:15:20	     [echo] ../../MoonshineSharedCore
    build	07-Dec-2021 13:15:20	     [echo] ../../MoonshineSharedCore/src/locale/{locale}
    build	07-Dec-2021 13:15:20	     [echo] ../src
    build	07-Dec-2021 13:15:20	    [mxmlc] Loading configuration file /Users/mmb/SDKs/Flex_4.16.1_AIR_33.1/frameworks/flex-config.xml
    build	07-Dec-2021 13:15:31	    [mxmlc] /Users/mmb/bamboo-agent-home/xml-data/build-dir/MS-BMOB-JOB1/Moonshine-IDE/ide/MoonshineWEBevolved/src/actionScripts/impls/IFileBridgeImp.as(39): col: 15 Error: Interface method get isBrowsed in namespace actionScripts.interfaces:IFileBridge not implemented by class actionScripts.impls:IFileBridgeImp.
    build	07-Dec-2021 13:15:31	    [mxmlc] 
    build	07-Dec-2021 13:15:31	    [mxmlc] 	public class IFileBridgeImp implements IFileBridge
    build	07-Dec-2021 13:15:31	    [mxmlc] 	             ^
    build	07-Dec-2021 13:15:31	    [mxmlc] 
    build	07-Dec-2021 13:15:32	    [mxmlc] /Users/mmb/bamboo-agent-home/xml-data/build-dir/MS-BMOB-JOB1/Moonshine-IDE/ide/MoonshineSharedCore/src/actionScripts/plugin/recentlyOpened/RecentlyOpenedPlugin.as(83): col: 32 Error: Access of undefined property CreateProject.
    build	07-Dec-2021 13:15:32	    [mxmlc] 
    build	07-Dec-2021 13:15:32	    [mxmlc] 			dispatcher.addEventListener(CreateProject.EVENT_SAVE_PROJECT_CREATION_FOLDERS, onNewProjectPathBrowse, false, 0, true);
    build	07-Dec-2021 13:15:32	    [mxmlc] 			                            ^
    build	07-Dec-2021 13:15:32	    [mxmlc] 
    build	07-Dec-2021 13:15:32	    [mxmlc] /Users/mmb/bamboo-agent-home/xml-data/build-dir/MS-BMOB-JOB1/Moonshine-IDE/ide/MoonshineSharedCore/src/actionScripts/plugin/recentlyOpened/RecentlyOpenedPlugin.as(21): col: 41 Error: Definition actionScripts.plugins.as3project:CreateProject could not be found.
    build	07-Dec-2021 13:15:32	    [mxmlc] 
    build	07-Dec-2021 13:15:32	    [mxmlc] 	import actionScripts.plugins.as3project.CreateProject;
    build	07-Dec-2021 13:15:32	    [mxmlc] 	                                       ^
    build	07-Dec-2021 13:15:32	    [mxmlc] 
    error	07-Dec-2021 13:15:34	
    error	07-Dec-2021 13:15:34	BUILD FAILED
    error	07-Dec-2021 13:15:34	/Users/mmb/bamboo-agent-home/xml-data/build-dir/MS-BMOB-JOB1/Moonshine-IDE/ide/MoonshineSharedCore/build/build-lib.xml:249: The following error occurred while executing this line:

It looks like your branch was created after the update to [IFileBridge.as](https://github.com/Moonshine-IDE/Moonshine-IDE/commit/731850e4f9313cfa41ccc480abc8afafe3b78ace#diff-da46675aad6696c6101389b1937f2d561624029b0edc7bf0e7464b07b6854fcc), but before the corresponding fix for [IFileBridgeImpl.as](https://github.com/Moonshine-IDE/Moonshine-IDE/commit/b907ffb4affb228ab03247174de94fe7961cb291#diff-f3853c4e175b82f0f31f489d1ac6af11ce5bdec4b129e66939385582fde7cf5b).  Please merge master in to the branch when you have a chance.

The MoonshineDESKTOPevolved build still works, though, so I did a quick test.  
1. Create a new Domino Visual Editor project
2. Add a simple label to the form.  
3. Project > Generate Apache Royale Project. See notes below
4. Set a project name and change the SDK to Royale
5. Project > Build & Run
6. BUG:  The project builds, but Moonshine tries the launch the wrong project.
![image](https://user-images.githubusercontent.com/24532731/145117713-efe07a1a-0805-4fa6-aff6-7289e1edeec1.png)


Investigating the launch bug, I found that you are using this setting:

  <jsOutput>
    <option path="bin"/>
  </jsOutput>

But this is set like this for a new Royale project with the default template:

  <jsOutput>
    <option path="/Users/-user-/Documents/Flex/temp_workspace/TestRoyaleJewel_1206"/>
  </jsOutput>

It seems that if we use a relative path here, it will use the last selected path from the filechooser.  If I change this to an absolute path for the generated project, this resolves the launch bug.  For now, try to match the logic for the normal Royale project, but we may revisit this later as we try to make projects easier to share.

The New Project form looked like this:
![image](https://user-images.githubusercontent.com/24532731/145117578-e8a41d4b-b9e1-452a-9c39-717f287d3a85.png)

Requested Changes:
- Select Apache Royale SDK by default (i.e. Apache Royale 0.9.9-SNAPSHOT).  See the existing form for Royale projects for this logic.
- "Porject" should be "Project"
- Change the Project Name and Template Type (see below) to "Apache Royale Project REST".  This generated project is not a Visual Editor project, so I think the current name is confusing if we want users to be able to create this project directly.

I also noticed a duplicated "Select Template Type" field in the Royale Browser Project.  This should be added to the existing Select Template Type combobox.
![image](https://user-images.githubusercontent.com/24532731/145117605-c3670ef2-b3a8-40fd-bc7c-fcc318c1e4c1.png)


I noticed that the generated project has the `visualeditor-src` directory.  I think this is not from the generated project, so it should be removed.

**Comment by JoelProminic** [comment]
@rat-moonshine made some updates to the template in the branch for #704.  You can find his templates here:  https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates/royaleTabularCRUD

**Comment by JoelProminic** [comment]
@feather812002, I saw that [this commit in VisualEditorConverterLib](https://github.com/Moonshine-IDE/VisualEditorConverterLib/commit/16db2d695d8e225269cb38ccb1c33abed4ddd939) broke the main Moonshine builds:

    build	10-Dec-2021 11:35:29	build:
    build	10-Dec-2021 11:35:29	     [echo] Building Visual Editor swc
    build	10-Dec-2021 11:35:30	    [compc] Loading configuration file /Users/-user-/SDKs/Flex_4.16.1_AIR_33.1/frameworks/air-config.xml
    build	10-Dec-2021 11:35:30	    [compc] Loading configuration file /Users/-user-/bamboo-agent-home/xml-data/build-dir/MS-MGOSX-JOB1/MockupVisualEditor/compile-config.xml
    build	10-Dec-2021 11:35:39	    [compc] .../MockupVisualEditor/src/view/domino/surfaceComponents/components/DominoTabView.as(104): col: 38 Error: Implicit coercion of a value of type components.domino:DominoTabView to an unrelated type interfaces.components:ITabView.
    build	10-Dec-2021 11:35:39	    [compc] 
    build	10-Dec-2021 11:35:39	    [compc] 			component = new components.domino.DominoTabView(this);
    build	10-Dec-2021 11:35:39	    [compc] 			                                  ^
    build	10-Dec-2021 11:35:39	    [compc] 
    build	10-Dec-2021 11:35:39	    [compc] .../MockupVisualEditor/src/defaults.css(120): Error: Invalid Embed directive in stylesheet - can't resolve source 'Embed("/../assets/images/shape_move_front.png")'.
    build	10-Dec-2021 11:35:39	    [compc] 
    build	10-Dec-2021 11:35:39	    [compc] 	icon: Embed("/../assets/images/shape_move_front.png");
    build	10-Dec-2021 11:35:39	    [compc] 

I reverted the change for now.  If you need to make more updates to VisualEditorConverterLib for this issue, please create a branch.

**Comment by JoelProminic** [comment]
@rat-moonshine got the #704 branch working again, and I wrote up some details here:  https://github.com/Moonshine-IDE/Moonshine-IDE/issues/704#issuecomment-991288458

One thing that was missing from this branch is that the data was not bound to a local model.  This is something I want to add for this issue in the long term, but for now we should focus on converting basic components into a reasonable Royale form.

@feather812002, talk to @rat-moonshine if you have any questions about the how this Royale code was generated.

**Comment by JoelProminic** [comment]
@feather812002 reported that he was having trouble with the new toRoyaleConvertCode method.

Context:
https://github.com/Moonshine-IDE/Moonshine-IDE/blob/40f700b90181896b468fd03fb8442cc2436accf0/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as#L989
https://github.com/Moonshine-IDE/MockupVisualEditor/blob/a4807fca456d0c51c150b77f07728c81afe582e1/src/utils/EditingSurfaceWriter.as#L267

The alert in the second link triggers, but the [toRoyaleConvertCode()](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/a4807fca456d0c51c150b77f07728c81afe582e1/src/utils/EditingSurfaceWriter.as#L274) method here returns null, or the `element` is null.

I tried to use the debugger for this today, but I found that I could not get the breakpoints to work with MockupVisualEditor.  @rat-moonshine reported that I needed to build the VisualEditor-0.2.0.swc in debug mode in order to do this.  I used the Ant script in MockupVisualEditor for this (update the target path as necessary).  

    ant -f build.xml -DMOONSHINE_LIBS_PATH=../Moonshine-IDE/ide/MoonshineDESKTOPevolved/libs/

Stepping through `EditingSurfaceWriter.toRoyaleCode`, I found that the only element in the container was a [MainApplication](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/a4807fca456d0c51c150b77f07728c81afe582e1/src/view/domino/surfaceComponents/components/MainApplication.as) instance.  I see this does not yet have a toRoyaleConvertCode method like I see for [DominoButton](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/a4807fca456d0c51c150b77f07728c81afe582e1/src/view/domino/surfaceComponents/components/DominoButton.as#L569)

For MainApplication, you can start by generating the Royale code for each of the children and combining them in the simplest way.  We'll need to update this more to handle layouts and other properties, but we should start with the easiest solution that will let us proceed.

**Comment by JoelProminic** [comment]
I confirmed that I can build a simple Royale form with a basic horizontal layout:
![image](https://user-images.githubusercontent.com/24532731/148299460-b5424cd3-74bf-43f8-8d5a-39b429a33d21.png)

However, I did notice that there was an extra nested bin directory for the output path:
![image](https://user-images.githubusercontent.com/24532731/148299533-a1ca11df-9cc0-4b43-9ffe-e0f70c5fde89.png)

It looks like we don't want the `bin` in the JavaScript Output Path.  If we need a value, we could try using "`.`" to indicate the current directory.

Generated project:
![image](https://user-images.githubusercontent.com/24532731/148299711-f477c8cd-b338-436e-a4a5-c29b1ee6e8af.png)

Normal Royale Project:
![image](https://user-images.githubusercontent.com/24532731/148299846-317127a2-a304-4486-9d1b-4854fa64b01f.png)

**Comment by piotrzarzycki21** [comment]
@feather812002 @rat-moonshine I'm getting following error during Moonshine source code build: 

```
MoonshineSharedCore/src/elements/templates/projects/Apache Royale Project REST/src/RoyaleTabularCRUDTemplate.mxml:4
[MoonshineSharedCore]: Could not resolve <j:Application> to a component implementation.
```

Why that particular template do not follow rule as any other by having name with extension *.template ?

**Comment by rat-moonshine** [comment]
> Why that particular template do not follow rule as any other by having name with extension *.template ?

Unfortunately, I'm not aware the templating in #675. However, this named similar to CRUD POC that I worked previously as #704 , and here was the old template for Royale that we were using - https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates/royaleTabularCRUD/project/src. I'm not sure about what rule you're referring to.

**Comment by piotrzarzycki21** [comment]
I was finally able to build and launch project from sources. I have made first small improvements. I will follow with reviewing code farther.

**Comment by piotrzarzycki21** [comment]
@feather812002 I'm currently working on some refactoring your part of the code. I will summarise changes once it will be ready. Whatever you are working on please make a small stop. These changes clean up a bit architecture which seems to me went out of control during work on this issue.

**Comment by JoelProminic** [comment]
There was an update today which broke the Bamboo builds.  @rat-moonshine fixed this for master, but the changes will need to be merged before the build for this branch works again.  This can wait for @piotrzarzycki21's refactoring, but I rely on this build for testing.

**Comment by piotrzarzycki21** [comment]
I have pushed changes to branches. Summary:
- Domino components were mixed with some classes and interfaces from PrimeFaces - I have decoupled them. After diving farther in the code I did understand that we are trying to have two type of outputs in domino components, instead one as it was with Flex and PrimeFaces. These additional output or any new one in future should be marked using some interfaces in this case I have created [IRoyaleComponentConverter](https://github.com/Moonshine-IDE/VisualEditorConverterLib/blob/features/issue_675_royale_generate_domino_visual_editor/src/interfaces/IRoyaleComponentConverter.as). If component implements it means that it outputting Royale code.
- I have moved [Div](https://github.com/Moonshine-IDE/VisualEditorConverterLib/blob/features/issue_675_royale_generate_domino_visual_editor/src/components/common/Div.as) and [RootDiv](https://github.com/Moonshine-IDE/VisualEditorConverterLib/blob/features/issue_675_royale_generate_domino_visual_editor/src/components/common/RootDiv.as) to separate common package cause it's being used by PrimeFaces and Domino
- I renamed IPrimeFacesSurfaceComponent to [IGetChildrenSurfaceComponent](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/features/issue_675_royale_generate_domino_visual_editor/src/view/interfaces/IGetChildrenSurfaceComponent.as) cause it's being used in both PrimeFaces and Domino
- I have removed all occurrences of Alert - @feather812002 if you are using it for debug purposes please do not commit it to the repository. Potential problems and errors should be handled in UI which is using both libraries not inside of them. 

I think after my changes everything should be tested if it's still working as it should be.

**Comment by piotrzarzycki21** [comment]
@feather812002 @JoelProminic what actually is that Image means added in comment for method nameAttribute in class [DominoInputText](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/features/issue_675_royale_generate_domino_visual_editor/src/view/domino/surfaceComponents/components/DominoInputText.as)

**Comment by piotrzarzycki21** [comment]
@rat-moonshine could you explain what actually [this means](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/a4a503a3fd79888f97b38b5ccb97927188f15d3b/ide/MoonshineSharedCore/src/elements/templates/projects/Apache%20Royale%20Project%20REST/src/view/MainContent.mxml#L26) in [Apache Royale REST](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/features/issue_675_royale_generate_domino_visual_editor/ide/MoonshineSharedCore/src/elements/templates/projects/Apache%20Royale%20Project%20REST/) template ? 

@JoelProminic after exporting Domino form to Royale - does project previously build ? 

Apache Royale REST template was suppose to be used to different purpose than this ? Should I create separate simple template for this project ?

**Comment by rat-moonshine** [comment]
> @rat-moonshine could you explain what actually this means in Apache Royale REST template ?

I think these are placeholders which replaced with corresponding information by Moonshine while generating a full-Royale project. 

Although, I'm not aware how the templates generates contents for the Royale project here, but it seems like similar approach that I followed in #704 .

There, to generate the template Royale project's content based on the OnDisk project elements, we had placeholders in different files, i.e. https://github.com/Moonshine-IDE/Moonshine-IDE/blob/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates/royaleTabularCRUD/project/src/views/MainContent.mxml.template#L30. Or, the element templates with different placeholders to replace with appropriate information here - https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates/royaleTabularCRUD/elements.

**Comment by piotrzarzycki21** [comment]
It sounds to me this project should rather have separate template.

**Comment by rat-moonshine** [comment]
In #704 it had all its separate templating area with corresponding elements - https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates/royaleTabularCRUD.

**Comment by piotrzarzycki21** [comment]
> In #704 it had all its separate templating area with corresponding elements - https://github.com/Moonshine-IDE/Moonshine-IDE/tree/features/issue_704_royale_CRUD_tabular_interface/ide/MoonshineSharedCore/src/elements/templates/royaleTabularCRUD.

Ok but it is for that specific purpose. I think this issue which I'm trying to help with should use different template not share code with #704

**Comment by JoelProminic** [comment]
> @JoelProminic after exporting Domino form to Royale - does project previously build ?

Yes, I could generate, build, and run the Royale project.  However, only a couple of components were implemented, so the results are not very interesting so far.

I pointed @feather812002 at the template from #704 as an existing template that we could use.  I'm fine with switching to different templates if you have a better design, though.  We are still early in the Royale conversion process.

**Comment by feather812002** [comment]
> @feather812002 @JoelProminic what actually is that Image means added in comment for method nameAttribute in class [DominoInputText](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/features/issue_675_royale_generate_domino_visual_editor/src/view/domino/surfaceComponents/components/DominoInputText.as)?

The image code require from Justin,
So, that when we generated the ASDoc it will contain a image in the help document.

**Comment by JoelProminic** [comment]
Unfortunately, Base64 encoded image for DominoInputText prevents us from opening the file normally in the GitHub interface.  It is also a giant block of data to scroll through in other editors.  We need to find a different solution for this.

One option would be to upload the images on a website (even with the image interface for GitHub) and include them this way.  If we want to keep these images in GitHub, then we could download the images directly from the repository URL.

In JavaDoc, I see we can [include images](https://stackoverflow.com/questions/10779169/including-images-in-javadocs) with the `doc-files` directory.

I see @feather812002 tried to add images like this for [MockupVisualEditor](https://github.com/Moonshine-IDE/MockupVisualEditor/tree/master/doc-files), but I see this didn't show up in the compiled documentation  We may be missing an option for this in the asdoc documentation, though.

**Comment by JoelProminic** [comment]
On a separate note, I see that @feather812002's branch has been failing in Bamboo since [this commit](https://github.com/Moonshine-IDE/VisualEditorConverterLib/commit/624418993bca4ace2a5579ebc4e468b962d240e5) from yesterday:

    build	02-Feb-2022 06:50:03	build:
    build	02-Feb-2022 06:50:03	     [echo] Building VisualEditorConverterLib swc
    build	02-Feb-2022 06:50:03	     [echo] VisualEditorConverterLib-0.1.0.swc
    build	02-Feb-2022 06:50:03	    [compc] Loading configuration file /Users/-user-/SDKs/Flex_4.16.1_AIR_33.1/frameworks/flex-config.xml
    build	02-Feb-2022 06:50:03	    [compc] Loading configuration file -snip-/MS-BMOB-JOB1/VisualEditorConverterLib/compile-config.xml
    build	02-Feb-2022 06:50:08	    [compc] -snip-/MS-BMOB-JOB1/VisualEditorConverterLib/src/components/domino/DominoTabView.as(207): col: 69 Error: Access of undefined property Div.
    build	02-Feb-2022 06:50:08	    [compc] 
    build	02-Feb-2022 06:50:08	    [compc] 						var internalDiv:Object = dominoconv.getNewInstanceOfComponent(Div.ELEMENT_NAME);
    build	02-Feb-2022 06:50:08	    [compc] 						                                                              ^
    build	02-Feb-2022 06:50:08	    [compc] 
    build	02-Feb-2022 06:50:08	    [compc] -snip-/MS-BMOB-JOB1/VisualEditorConverterLib/src/components/domino/DominoTabView.as(469): col: 70 Error: Access of undefined property Div.
    build	02-Feb-2022 06:50:08	    [compc] 
    build	02-Feb-2022 06:50:08	    [compc]                     container = dominoconv.getNewInstanceOfComponent(Div.ELEMENT_NAME);
    build	02-Feb-2022 06:50:08	    [compc]                                                                      ^
    build	02-Feb-2022 06:50:08	    [compc] 
    build	02-Feb-2022 06:50:08	    [compc] -snip-/MS-BMOB-JOB1/VisualEditorConverterLib/src/components/domino/DominoTabView.as(26): col: 30 Error: Definition components.primeFaces:Div could not be found.
    build	02-Feb-2022 06:50:08	    [compc] 
    build	02-Feb-2022 06:50:08	    [compc] 	import components.primeFaces.Div;
    build	02-Feb-2022 06:50:08	    [compc] 	                            ^
    build	02-Feb-2022 06:50:08	    [compc] 
    error	02-Feb-2022 06:50:08	
    error	02-Feb-2022 06:50:08	BUILD FAILED
    error	02-Feb-2022 06:50:08	-snip-/MS-BMOB-JOB1/VisualEditorConverterLib/build.xml:39: compc task failed
    error	02-Feb-2022 06:50:08

**Comment by piotrzarzycki21** [comment]
@feather812002 @JoelProminic @rat-moonshine I have been working on applying new template on this branch. After many hours of fight I have bump into an issue where it looks like creating "Royale browser project" seems to be completely broken.  I was working on that I didn't realise until I tried create "Royale Browser project" on development build and the one from this branch. 

Development build:
<img width="1509" alt="Screenshot 2022-02-28 at 11 04 15" src="https://user-images.githubusercontent.com/24554795/155966393-bc0fee59-94b7-43de-b4ba-6754e76124de.png">

Branch build:
<img width="1492" alt="Screenshot 2022-02-28 at 11 03 39" src="https://user-images.githubusercontent.com/24554795/155966448-b6306ded-a11c-44db-9d32-e215d0f4ff7f.png">

On branch build we should have same look as it is on development. It looks like @feather812002 when you were working on it you didn't take care of scenerio with simple project creation for Royale ? 
I'm not sure and I don't know how to fix that right now, so any help is appreciated.

**Comment by rat-moonshine** [comment]
Why it has `Select Template Type` and `Select Template Type1` @feather812002 ?

**Comment by feather812002** [comment]
> Why it has `Select Template Type` and `Select Template Type1` @feather812002 ?
@rat-moonshine 
Hello, Where are you found it ?

I work on other brach at now , So, I can't search the Select Template Type and Select Template Type1 string from Project.

**Comment by rat-moonshine** [comment]
@feather812002 the labels are from second image that @piotrzarzycki21 posted [here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/675#issuecomment-1054109330).

**Comment by piotrzarzycki21** [comment]
I will try to nailed it down. I will review commits, maybe revert something to get things working as it was before and next try to apply implementation with simpler template. It's already took me days to figure out things and now it gets worse a bit.

**Comment by feather812002** [comment]
@rat-moonshine  and @piotrzarzycki21 

Okay, I review the code into details on the branch.
I think we all have some misunderstand , I will clear these "Select Template Type"  and "Select Template Type1"

First we have two options from Fiel->New
1. Royale Browser project
2. Visual Editor project.

From my get require from @JoelProminic and @JustinProminic 
I think I need work around the visual royale UI project , 
Other word we need some visual royale UI components work with Moonshine-IDE Visual Edtior.
So, I think it is different with Royale Browser project, 
This is why I add a new  "Select Template Type1" into  "File->New->Visual Editor project"

@piotrzarzycki21  you post the first image is from "File->New->Royale Browser project "
I check this with the branch , it full same with your post.

The seconde post image ("Select Template Type1") this is only for Visuale Royale project test.

The branch not end ,right ? so ,I add this only for test.

and  because the "Royale Browser project," is different with "Visuale Royale project", so I add this into "File->New->Visual Editor project", it work on other Royale template for this new visual Royale project.

I will back to this task after I done other issue, and try let it work more close master version .
Now this is just test branch UI options.

So,any other question ?

**Comment by piotrzarzycki21** [comment]
@feather812002 The only thing which you need to check on branch is why when I do "File->New->Royale Browser project " I'm not getting following view:
<img width="1504" alt="Screenshot 2022-03-01 at 16 04 06" src="https://user-images.githubusercontent.com/24554795/156193541-dc4dfda9-05a9-4e74-8b0b-0cb6df9e0576.png">

This is what I should get when I'm creating new Royale project. Once you fix this issue and commit to this branch - stop work, cause I will change template which is now used for exporting Visual Editor Project to Royale - Once I do this we will decide internally who will be progressing farther with above work. 

Let me know if everything is clear to you.

**Comment by feather812002** [comment]
@piotrzarzycki21 

1. Okay ,I fix this , Now I confirm it work same with your post:
![image](https://user-images.githubusercontent.com/8543391/156352125-a69a230b-f00c-48c7-86cc-e2ce1c68c0bb.png)
2. I can't commit it to Github ,Looks like you removed my commit permissions , right ?
Looks like you removed my commit permissions
![image](https://user-images.githubusercontent.com/8543391/156352721-31ecfebc-0985-4d30-8560-52526113c1a3.png)

So, I no way commit my update until now.

3. I already have a long time not work on this . So, you can free arrange for anyone to continue this project.

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21
> 
> 1. Okay ,I fix this , Now I confirm it work same with your post:
>    ![image](https://user-images.githubusercontent.com/8543391/156352125-a69a230b-f00c-48c7-86cc-e2ce1c68c0bb.png)
> 2. I can't commit it to Github ,Looks like you removed my commit permissions , right ?
>    ![image](https://user-images.githubusercontent.com/8543391/156353308-c25f94e7-e439-43a1-ac10-2fee47f1bb3b.png)
> 
> So, I no way commit my update until now.
> 
> 3. I already have a long time not work on this . So, you can free arrange for anyone to continue this project.

I didn't remove any of your permissions. Pull changes and try again. @JoelProminic could you verify what is happening ?

**Comment by feather812002** [comment]
> 

Okay , @piotrzarzycki21 
I still can't push it Git after I pull it some times from last update on this branch.

So, I will post what is my change , it will easy , you just need commet :
https://github.com/Moonshine-IDE/Moonshine-IDE/blob/3e84808e492fa6f368563bbfe1284e5ce968c2d2/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as#L494

from line 494 - 502, it will back old UI that you post on above.

I will discuss with Joel on the tonight ,why I can't push the update to Git

Thanks.

**Comment by piotrzarzycki21** [comment]
> > 
> 
> Okay , @piotrzarzycki21 I still can't push it Git after I pull it some times from last update on this branch.
> 
> So, I will post what is my change , it will easy , you just need commet :
> 
> https://github.com/Moonshine-IDE/Moonshine-IDE/blob/3e84808e492fa6f368563bbfe1284e5ce968c2d2/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as#L494
> 
> from line 494 - 502, it will back old UI that you post on above.
> 
> I will discuss with Joel on the tonight ,why I can't push the update to Git
> 
> Thanks.

I will wait for your commit. This issue with git need to be resolved.

**Comment by rat-moonshine** [comment]
> @piotrzarzycki21
> 
> 1. Okay ,I fix this , Now I confirm it work same with your post:
>    ![image](https://user-images.githubusercontent.com/8543391/156352125-a69a230b-f00c-48c7-86cc-e2ce1c68c0bb.png)
> 2. I can't commit it to Github ,Looks like you removed my commit permissions , right ?
>    ![image](https://user-images.githubusercontent.com/8543391/156353308-c25f94e7-e439-43a1-ac10-2fee47f1bb3b.png)
> 
> So, I no way commit my update until now.
> 
> 3. I already have a long time not work on this . So, you can free arrange for anyone to continue this project.

Based on the error screenshot reported by @feather812002 , it seems he's local branch is behind from remote:

"hint: Updates were rejected because a pushed branch tip is behind its remote"

As the _hints_ suggested, try to do a `git pull` first, then try a `push`.

**Comment by feather812002** [comment]
okay, @rat-moonshine and @piotrzarzycki21 
You are correct, my bad.
After I pull and I can commit it

**Comment by piotrzarzycki21** [comment]
@feather812002 I think something is still wrong after your changes. Should generation of new project look like that when I do menu **Project** -> **Generate Apache Royale Project** ?

<img width="1493" alt="Screenshot 2022-03-03 at 12 49 20" src="https://user-images.githubusercontent.com/24554795/156559473-5b5a1e3c-e440-479e-8371-392a3eeb1347.png">

**Comment by feather812002** [comment]
@piotrzarzycki21 
I work on latest update from you on : 
```
commit 732f7183e6c63a0f06ea714facbaaa6c52081372 (HEAD -> features/issue_675_royale_generate_domino_visual_editor, origin/features/issue_675_royale_generate_domino_visual_editor)
Author: Piotr Zarzycki <piotrzarzycki21@gmail.com>
```

I want repeat your repeat your bug that you report on above.

But when I build it with Moonshine-IDE I got follow error:
Copying resource: templates
```
: Error #2037
: Error: Error #2037
: 	at actionScripts.plugins.as3project.mxmlc::MXMLCPlugin/copyingResources()
: 	at actionScripts.plugins.as3project.mxmlc::MXMLCPlugin/shellData()
:
: Click here to Report a Bug
: Compilation of MoonshineDESKTOPevolved finished.
```

Any suggestion about it ?

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21 I work on latest update from you on :
> 
> ```
> commit 732f7183e6c63a0f06ea714facbaaa6c52081372 (HEAD -> features/issue_675_royale_generate_domino_visual_editor, origin/features/issue_675_royale_generate_domino_visual_editor)
> Author: Piotr Zarzycki <piotrzarzycki21@gmail.com>
> ```
> 
> I want repeat the bug that you report on above.
> 
> But when I build it with Moonshine-IDE I got follow error: Copying resource: templates
> 
> ```
> : Error #2037
> : Error: Error #2037
> : 	at actionScripts.plugins.as3project.mxmlc::MXMLCPlugin/copyingResources()
> : 	at actionScripts.plugins.as3project.mxmlc::MXMLCPlugin/shellData()
> :
> : Click here to Report a Bug
> : Compilation of MoonshineDESKTOPevolved finished.
> ```
> 
> Any suggestion about it ?

I was able build Moonshine from Intellij. I will setup Moonshine sources inside Moonshine on Monday, cause I'm assuming this is the case which you are presenting now...

**Comment by piotrzarzycki21** [comment]
@feather812002 I was able build and run Moonshine sources inside Moonshine.

**Comment by piotrzarzycki21** [comment]
> @feather812002 I think something is still wrong after your changes. Should generation of new project look like that when I do menu **Project** -> **Generate Apache Royale Project** ?
> 
> <img alt="Screenshot 2022-03-03 at 12 49 20" width="1493" src="https://user-images.githubusercontent.com/24554795/156559473-5b5a1e3c-e440-479e-8371-392a3eeb1347.png">

@feather812002 This issue is still relevant and I wold like you to take a priority in fixing it. Let me know if after latest pull you still have problems with build.

**Comment by feather812002** [comment]
@piotrzarzycki21  can you details what wrong on the UI interface ?
This is my test as follow :  
1) Project -> Generate Apache Royale Project
![image](https://user-images.githubusercontent.com/8543391/157054061-21815f76-0037-43b8-bff7-63d056705a6b.png)
2) File->New->Royale Browser project
![image](https://user-images.githubusercontent.com/8543391/157054267-e256fe9a-bb2a-49bb-a1f4-7f7b3a95ec37.png)
They have same UI, 

So, what wrong on the UI from click Project -> Generate Apache Royale Project?

**Comment by feather812002** [comment]
@piotrzarzycki21 I work around this at now ,it need some time , when it ready ,I will let you know.

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21 I work around this at now ,it need some time , when it ready ,I will let you know.

Are you going commit that change ? I would like to work on adding simplified template to exported project

**Comment by feather812002** [comment]
> > @piotrzarzycki21 I work around this at now ,it need some time , when it ready ,I will let you know.
> 
> Are you going commit that change ? I would like to work on adding simplified template to exported project

No, I not yet commit any update until now , I will need check the code and find what code show the UI in this style ,then I can fix it follow your require .

You can add anything ,before I commit it , I will pull your update.

**Comment by feather812002** [comment]
![image](https://user-images.githubusercontent.com/8543391/157235750-a9111f3e-4385-4415-9945-691933e06cc8.png)
@piotrzarzycki21 please check this UI meet your require.
If it is okay , I will commit it ,if you think it still need improve ,let me know .

**Comment by piotrzarzycki21** [comment]
Branch has been merged to master

**Comment by JoelProminic** [comment]
I see `Project > Generate Apache Royale Project` is visible now in the nightly build.   I tested it with the project I have been using for NSF Converter testing, but I ended up with a bunch of errors.  However, this is a pretty complicated database, so I am not surprised that I got errors when this feature is not complete.

If we are not closer to completion by the time we are ready for the 3.3.2 release, we should disable this menu item.

**Comment by piotrzarzycki21** [comment]
> I see `Project > Generate Apache Royale Project` is visible now in the nightly build. I tested it with the project I have been using for NSF Converter testing, but I ended up with a bunch of errors. However, this is a pretty complicated database, so I am not surprised that I got errors when this feature is not complete.
> 
> If we are not closer to completion by the time we are ready for the 3.3.2 release, we should disable this menu item.

I have removed temporary that option to do not confuse user with disable.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine How can I make one of the option of menu make available only for a specific type of project. Example if I have selected Apache Royale project - in menu "Project" I would like to remove one option, but when project will be switched to different type for example Domino Visual Editor - this option should appear. Where should I look ? What determine that behavior ?

**Comment by rat-moonshine** [comment]
Are you asking something similar to https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/ui/menu/ProjectMenu.as ? The `Project` menu re-generates on tree-selection change between different projects.

**Comment by piotrzarzycki21** [comment]
I was able finally add new template, but it doesn't pickup yet Domino Visual Editor code yet to have transformation to Royale. I went farther into initial logic from @feather812002 and it seems to me that approach of doing transformation is being made based on opened and active currently editor. I need to change that to based fully on file itself regardless it's opened or not.

**Comment by rat-moonshine** [comment]
> Yes Exactly I'm asking about that - Where should I look, what should I do ?

https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/ui/menu/ProjectMenu.as#L66 executes every time user change selection in sidebar and `model.activeProject` updates. You possibly need to find your targeted project-type in the method, i.e. Royale, Visual Editor etc. and do modification in their menu-generation method, i.e. https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/ui/menu/ProjectMenu.as#L170.

Beside this, I'm not sure if you're looking for some other information than this basic.

**Comment by piotrzarzycki21** [comment]
@JoelProminic , @feather812002, @rat-moonshine I have made one significant changed in DominoCoverter class - If you will see any kind of problems with Tab code and proper translation to XML let me know.

**Comment by piotrzarzycki21** [comment]
While working on convertion to Royale I have bump into an architectural issue where classes responsible for conversion where being treated and made as singletons. We have made long discussion and session with @Aszusz over that and decided get rid of singletons in favor of having these classes static - It cleans up tons of code and made these classes easier to use - made them more cleaner. 

- [DominoConverter](https://github.com/Moonshine-IDE/VisualEditorConverterLib/blob/master/src/converter/DominoConverter.as) and [PrimeFacesConverter](https://github.com/Moonshine-IDE/VisualEditorConverterLib/blob/master/src/converter/PrimeFacesConverter.as) are now fully static - they contains methods, but do not hold any state. 
- Lookup object which has been passed to converters now is being splitted into two categories UI component providers ([FlexUILookup](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/master/src/lookup/FlexUILookup.as), [PrimeFacesUILookup](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/master/src/lookup/PrimeFacesUILookup.as), [DominoUILookup](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/master/src/lookup/DominoUILookup.as)) and non-UI component providers ([FlexNonUILookup](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/master/src/lookup/FlexNonUILookup.as), [PrimeFacesNonUILookup](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/master/src/lookup/PrimeFacesNonUILookup.as), [DominoNonUILookup](https://github.com/Moonshine-IDE/MockupVisualEditor/blob/master/src/lookup/DominoNonUILookup.as)). Mockup visual editor need to "decide" what Lookup use. 

@feather812002 provide me all branches which you are actively working on for your issue from Moonshine-IDE, MockupVisualEditor and VisualEditorConverterLib repository - I would like to merge these changes to your branches myself.

**Comment by rat-moonshine** [comment]
A `text="{...}"` could be safe to compiler for this (?)

**Comment by piotrzarzycki21** [comment]
Actually issue is completely different. Some labels after conversion has properties inside and compiler complaining that properties doesn't exists. It is correct they are not declared anywhere - I'm not sure what to do with that yet. In below error compiler points into "disk"

```
HostConfigSnapshot.mxml(741): col: 155 Error: Access of possibly undefined property disk.
: <j:Label text="for disk in $(/opt/ooce/smartmontools/sbin/smartctl --scan | awk &amp;apos;{print $1}&amp;apos;); do printf &amp;quot;${disk} &amp;quot;; pfexec /opt/ooce/smartmontools/sbin/smartctl -H ${disk} | sed -n &amp;apos;/^=== START OF READ SMART DATA SECTION ===.*/ {N;s/.*:[[:space:]]*\([[:alnum:]]*\)/\1/p;D;}&amp;apos;; done" className="cursor-pointer" multiline="true"/>
```

**Comment by piotrzarzycki21** [comment]
@JoelProminic We were able fix issue with displaying commands in Labels in Jewel. Right now everything is working fine. Project which you have sent me is being converted fully to Royale - you can test it, but you have to remember that one file called "Config Value" has to be renamed to the version without spaces. In folders `/nsfs/nsf-moonshine/odp/Forms` and `visualeditor-src/main/webapp`.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I didn't put a notes. Do you remember if priority was in that order:
1) Make forms read only on demand
2) Put text and rich text editor into new line
3) Agreed on behaviour of paragraph

**Comment by JoelProminic** [comment]
The priorities I noted from yesterday's meeting were:
- fix missing newlines between labels and fields.
- Edit Mode/Read only

For the newline issue, I see that some of these cases have the command label and the field in the same paragraph (i.e. `hostname`), while others have separate paragraphs (i.e. `uname -a`):

![image](https://user-images.githubusercontent.com/24532731/169362499-c7943589-ae4b-4e4c-a6ac-c628f89ab92b.png)

I consider the same paragraph cases a bug in the conversion process.  I'll create a separate issue for this after I retest.

We'll need to clarify the definition of the Paragraph component for the separate paragraphs case.   We could either add clarification that Paragraph always indicates a separate line, or we can update the converted database to use vertical layout for the parent.  I'll keep you updated on this.

For the read-only and edit modes, I see two common use cases:
- We want the interface to be read-only.  This is case for OmniOS Performance Stats
- We want a full CRUD interface for the application.

We can brainstorm how we want to support these two cases.  
- Separate options for read-only and CRUD interfaces?  Would this result in more duplication?
- Configurable options for the forms for read-only/editable.  I'm not sure if this fits with the current design, though.
- Would it make sense to always generate the CRUD interface and then provide flags/instructions to disable the read interface?  This could be a unified solution if we aren't concerned that it will generate messy code.

We could handle the read-only and edit modes like we do for the MyDetails section in MyAccount.  This logic seems clean on the UI side, but I'm not sure how well it would work for generated code.

**Comment by Aszusz** [comment]
I checked the weird wrapping in the long commands:
![obraz](https://user-images.githubusercontent.com/7197014/169380022-bb6d6ad8-fbb8-4e50-a868-18333f610127.png)
and it turns out, there's `\n` character in the original forms, so it's not a conversion issue

**Comment by JoelProminic** [comment]
Thanks @Aszusz, I didn't see this when I was comparing the interfaces yesterday.  There will probably be other backslashes in the commands, so watch the escaping on this.  I can see some other examples in your screenshot:
![image](https://user-images.githubusercontent.com/24532731/169385214-812737bc-a22e-4b83-ac71-b5ffe49bfd40.png)

This is an unusual case that I wouldn't expect to see in other database, but this is one of the key databases we want to convert.

**Comment by piotrzarzycki21** [comment]
I have pushed changes related to "readonly" form. The changes I made is that every form has property which allows switch between readonly/edit mode. Once you convert domino project to Royale and open any view you will see inside property called `isDisabled`, default value is `false` - which means form is in edit mode - change it to `true` and recompile application. 

<img width="1792" alt="Screenshot 2022-05-23 at 11 13 22" src="https://user-images.githubusercontent.com/24554795/169786133-0ee9129b-c6f7-4f6c-a781-91ec8ee8171a.png">

However I'm not sure to which direction I should go from that point. I see two options:
1) User on it's own provides some Button/Switch for his form which changes in individual form property `isDisabled` to `true/false`
2) I'm adding during conversion such button, but I see in this action some inconsistency - cause this kind of button won't be something which is present in DXL. It's position may not be determined correctly. 

Thoughts ?

**Comment by piotrzarzycki21** [comment]
> I consider the same paragraph cases a bug in the conversion process. I'll create a separate issue for this after I retest.

@JoelProminic just to be clear this code is being considered as a bug ? Cause it translates in Moonshine as a one paragraph, but it should be two actually ? 

```
    <par def="1007" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
        <run>
          <font color="system" style="normal"/>Key:  
        </run>
        <field useappletinbrowser="false" allowtabout="false" defaultfocus="false" protected="false" sign="false" storelocally="false" value="Key" type="text" kind="editable" computeaftervalidation="false" allowmultivalues="false" width="100pt" height="30pt" bgcolor="#ffffff" name="Key"/>
      </par>
```

I need to wait till you fix that ? 

Currently we have following situation after conversion to Royale:

```
<VGroup> - Parent
     <HGroup> - Paragraph
     </HGroup>
</VGroup>
```

Each new paragraph is going to the new line. Is this how I should preserve that ?

**Comment by JoelProminic** [comment]
@piotrzarzycki21, your first example (from "ConfigValue") is correct as-is.  The label and the field are on the saem line in this case.  The layout should be like:

```
<VGroup> - Parent
     <HGroup> - Paragraph
        Label
        Field
     </HGroup>
</VGroup>
```

The problem is this case, from Host Config Snapshot:

```
                  <par def="1307" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
                    <run>
                      <font color="system" style="normal" name="monospace"/>$ hostname
                    </run>
                    <break/>
                    <run>
                      <font color="black" size="10pt" style="normal" name="sans-serif" pitch="variable" familyid="Roman" truetype="true"/><field listdisplayseparator="newline" listinputseparators="comma semicolon newline" useappletinbrowser="false" allowtabout="false" defaultfocus="false" storelocally="false" value="HostnameOutput" type="text" kind="editable" computeaftervalidation="false" allowmultivalues="false" width="106pt" height="30pt" bgcolor="#ffffff" name="HostnameOutput"/>
                    </run>
                  </par>
```

This example only has a single `<par>`, but it displays on two lines because of the `<break/>`.  I'm not sure why Domino Designer did this for some of the cases, but I'd like for NSFConverter to normalize this as two paragraphs in the intermediate XML.

See #1033 for more details.  In the meantime, you can treat the single `<par>` as a single line, so the layout will look the same as your example.  `<break/>` doesn't appear in the Mockup (and i don't think it makes sense to have it) so I don't want to support it.
